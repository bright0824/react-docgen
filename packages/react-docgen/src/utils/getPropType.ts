import type { NodePath } from '@babel/traverse';
import { getDocblock } from '../utils/docblock.js';
import getMembers from './getMembers.js';
import getPropertyName from './getPropertyName.js';
import isRequiredPropType from '../utils/isRequiredPropType.js';
import printValue from './printValue.js';
import resolveToValue from './resolveToValue.js';
import resolveObjectKeysToArray from './resolveObjectKeysToArray.js';
import resolveObjectValuesToArray from './resolveObjectValuesToArray.js';
import type { PropTypeDescriptor } from '../Documentation.js';
import type { ArrayExpression, Expression, SpreadElement } from '@babel/types';

function getEnumValuesFromArrayExpression(
  path: NodePath<ArrayExpression>,
): Array<Record<string, unknown>> {
  const values: Array<Record<string, unknown>> = [];

  path.get('elements').forEach((elementPath) => {
    if (!elementPath.hasNode()) return;

    if (elementPath.isSpreadElement()) {
      const value = resolveToValue(elementPath.get('argument'));

      if (value.isArrayExpression()) {
        // if the SpreadElement resolved to an Array, add all their elements too
        return values.push(...getEnumValuesFromArrayExpression(value));
      } else {
        // otherwise we'll just print the SpreadElement itself
        return values.push({
          value: printValue(elementPath),
          computed: !elementPath.isLiteral(),
        });
      }
    }

    // try to resolve the array element to it's value
    const value = resolveToValue(elementPath as NodePath<Expression>);

    return values.push({
      value: printValue(value.isImportDeclaration() ? elementPath : value),
      computed: !value.isLiteral(),
    });
  });

  return values;
}

function getPropTypeOneOf(
  type: PropTypeDescriptor,
  argumentPath: NodePath,
): PropTypeDescriptor {
  const value = resolveToValue(argumentPath);

  if (value.isArrayExpression()) {
    type.value = getEnumValuesFromArrayExpression(value);
  } else {
    const objectValues =
      resolveObjectKeysToArray(value) || resolveObjectValuesToArray(value);

    if (objectValues) {
      type.value = objectValues.map((objectValue) => ({
        value: objectValue,
        computed: false,
      }));
    } else {
      // could not easily resolve to an Array, let's print the original value
      type.computed = true;
      type.value = printValue(argumentPath);
    }
  }

  return type;
}

function getPropTypeOneOfType(
  type: PropTypeDescriptor,
  argumentPath: NodePath,
): PropTypeDescriptor {
  if (argumentPath.isArrayExpression()) {
    type.value = argumentPath.get('elements').map((elementPath) => {
      if (!elementPath.hasNode()) return;
      const descriptor: PropTypeDescriptor = getPropType(elementPath);
      const docs = getDocblock(
        elementPath as NodePath<Expression | SpreadElement>,
      );

      if (docs) {
        descriptor.description = docs;
      }

      return descriptor;
    });
  }

  return type;
}

function getPropTypeArrayOf(
  type: PropTypeDescriptor,
  argumentPath: NodePath,
): PropTypeDescriptor {
  const docs = getDocblock(argumentPath);

  if (docs) {
    type.description = docs;
  }

  const subType = getPropType(argumentPath);

  // @ts-ignore
  if (subType.name !== 'unknown') {
    type.value = subType;
  }

  return type;
}

function getPropTypeObjectOf(
  type: PropTypeDescriptor,
  argumentPath: NodePath,
): PropTypeDescriptor {
  const docs = getDocblock(argumentPath);

  if (docs) {
    type.description = docs;
  }

  const subType = getPropType(argumentPath);

  // @ts-ignore
  if (subType.name !== 'unknown') {
    type.value = subType;
  }

  return type;
}

function getFirstArgument(path: NodePath): NodePath | undefined {
  let argument: NodePath | undefined;

  if (path.isCallExpression()) {
    argument = path.get('arguments')[0];
  } else {
    const members = getMembers(path, true);

    if (members[0] && members[0].argumentPaths[0]) {
      argument = members[0].argumentPaths[0];
    }
  }

  return argument;
}

function isCyclicReference(
  argument: NodePath,
  argumentPath: NodePath,
): boolean {
  return Boolean(argument && resolveToValue(argument) === argumentPath);
}

/**
 * Handles shape and exact prop types
 */
function getPropTypeShapish(
  type: PropTypeDescriptor,
  argumentPath: NodePath,
): PropTypeDescriptor {
  if (!argumentPath.isObjectExpression()) {
    argumentPath = resolveToValue(argumentPath);
  }

  if (argumentPath.isObjectExpression()) {
    const value: Record<string, PropTypeDescriptor> = {};
    let rawValue: string | undefined;

    argumentPath.get('properties').forEach((propertyPath) => {
      // We only handle ObjectProperty as there is nothing to handle for
      // SpreadElements and ObjectMethods
      if (propertyPath.isObjectProperty()) {
        const propertyName = getPropertyName(propertyPath);

        if (!propertyName) return;

        const valuePath = propertyPath.get('value');
        const argument = getFirstArgument(valuePath);

        // This indicates we have a cyclic reference in the shape
        // In this case we simply print the argument to shape and bail
        if (argument && isCyclicReference(argument, argumentPath)) {
          rawValue = printValue(argument);

          return;
        }

        const descriptor = getPropType(valuePath);
        const docs = getDocblock(propertyPath);

        if (docs) {
          descriptor.description = docs;
        }
        descriptor.required = isRequiredPropType(valuePath);
        value[propertyName] = descriptor;
      }
    });

    type.value = rawValue ?? value;
  }

  return type;
}

function getPropTypeInstanceOf(
  _type: PropTypeDescriptor,
  argumentPath: NodePath,
): PropTypeDescriptor {
  return {
    name: 'instanceOf',
    value: printValue(argumentPath),
  };
}

const simplePropTypes = [
  'array',
  'bool',
  'func',
  'number',
  'object',
  'string',
  'any',
  'element',
  'node',
  'symbol',
  'elementType',
] as const;

function isSimplePropType(
  name: string,
): name is (typeof simplePropTypes)[number] {
  return simplePropTypes.includes(name as (typeof simplePropTypes)[number]);
}

type PropTypeHandler = (
  type: PropTypeDescriptor,
  argumentPath: NodePath,
) => PropTypeDescriptor;

const propTypes = new Map<
  string,
  (argumentPath: NodePath | undefined) => PropTypeDescriptor
>([
  ['oneOf', callPropTypeHandler.bind(null, 'enum', getPropTypeOneOf)],
  ['oneOfType', callPropTypeHandler.bind(null, 'union', getPropTypeOneOfType)],
  [
    'instanceOf',
    callPropTypeHandler.bind(null, 'instanceOf', getPropTypeInstanceOf),
  ],
  ['arrayOf', callPropTypeHandler.bind(null, 'arrayOf', getPropTypeArrayOf)],
  ['objectOf', callPropTypeHandler.bind(null, 'objectOf', getPropTypeObjectOf)],
  ['shape', callPropTypeHandler.bind(null, 'shape', getPropTypeShapish)],
  ['exact', callPropTypeHandler.bind(null, 'exact', getPropTypeShapish)],
]);

function callPropTypeHandler(
  name: PropTypeDescriptor['name'],
  handler: PropTypeHandler,
  argumentPath: NodePath | undefined,
): PropTypeDescriptor {
  let type: PropTypeDescriptor = { name };

  if (argumentPath) {
    type = handler(type, argumentPath);
  }

  if (!type.value) {
    // If there is no argument then leave the value an empty string
    type.value = argumentPath ? printValue(argumentPath) : '';
    type.computed = true;
  }

  return type;
}

/**
 * Tries to identify the prop type by inspecting the path for known
 * prop type names. This method doesn't check whether the found type is actually
 * from React.PropTypes. It simply assumes that a match has the same meaning
 * as the React.PropTypes one.
 *
 * If there is no match, "custom" is returned.
 */
export default function getPropType(path: NodePath): PropTypeDescriptor {
  let descriptor: PropTypeDescriptor | null = null;

  getMembers(path, true).some((member) => {
    const memberPath = member.path;
    let name: string | null = null;

    if (memberPath.isStringLiteral()) {
      name = memberPath.node.value;
    } else if (memberPath.isIdentifier() && !member.computed) {
      name = memberPath.node.name;
    }
    if (name) {
      if (isSimplePropType(name)) {
        descriptor = { name };

        return true;
      }
      const propTypeHandler = propTypes.get(name);

      if (propTypeHandler) {
        descriptor = propTypeHandler(member.argumentPaths[0]);

        return true;
      }
    }

    return;
  });

  if (descriptor) {
    return descriptor;
  }

  if (path.isIdentifier() && isSimplePropType(path.node.name)) {
    return { name: path.node.name };
  }

  if (path.isCallExpression()) {
    const callee = path.get('callee');

    if (callee.isIdentifier()) {
      const propTypeHandler = propTypes.get(callee.node.name);

      if (propTypeHandler) {
        return propTypeHandler(path.get('arguments')[0]);
      }
    }
  }

  return { name: 'custom', raw: printValue(path) };
}
