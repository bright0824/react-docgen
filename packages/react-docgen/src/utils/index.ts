import * as docblock from './docblock';
import * as expressionTo from './expressionTo';
import * as flowUtilityTypes from './flowUtilityTypes';
import * as traverse from './traverse';

export { docblock, expressionTo, flowUtilityTypes, traverse };

export { default as getClassMemberValuePath } from './getClassMemberValuePath';
export { default as getFlowType } from './getFlowType';
export { default as getMemberExpressionRoot } from './getMemberExpressionRoot';
export { default as getMemberExpressionValuePath } from './getMemberExpressionValuePath';
export { default as getMembers } from './getMembers';
export {
  default as getMemberValuePath,
  isSupportedDefinitionType,
} from './getMemberValuePath';
export { default as getMethodDocumentation } from './getMethodDocumentation';
export type { MethodNodePath } from './getMethodDocumentation';
export { default as getNameOrValue } from './getNameOrValue';
export { default as getParameterName } from './getParameterName';
export { default as getPropertyName, COMPUTED_PREFIX } from './getPropertyName';
export { default as getPropertyValuePath } from './getPropertyValuePath';
export { default as getPropType } from './getPropType';
export { default as getTSType } from './getTSType';
export { default as getTypeAnnotation } from './getTypeAnnotation';
export {
  default as getTypeFromReactComponent,
  applyToTypeProperties,
} from './getTypeFromReactComponent';
export { default as getTypeIdentifier } from './getTypeIdentifier';
export { default as getTypeParameters } from './getTypeParameters';
export type { TypeParameters } from './getTypeParameters';
export { default as isDestructuringAssignment } from './isDestructuringAssignment';
export { default as isExportsOrModuleAssignment } from './isExportsOrModuleAssignment';
export { default as isReactBuiltinCall } from './isReactBuiltinCall';
export { default as isReactChildrenElementCall } from './isReactChildrenElementCall';
export { default as isReactCloneElementCall } from './isReactCloneElementCall';
export { default as isReactComponentClass } from './isReactComponentClass';
export { default as isReactComponentMethod } from './isReactComponentMethod';
export { default as isReactCreateClassCall } from './isReactCreateClassCall';
export { default as isReactCreateElementCall } from './isReactCreateElementCall';
export { default as isReactForwardRefCall } from './isReactForwardRefCall';
export { default as isReactModuleName } from './isReactModuleName';
export { default as isRequiredPropType } from './isRequiredPropType';
export { default as isStatelessComponent } from './isStatelessComponent';
export { default as isUnreachableFlowType } from './isUnreachableFlowType';
export { default as match } from './match';
export { default as normalizeClassDefinition } from './normalizeClassDefinition';
export { default as parseJsDoc } from './parseJsDoc';
export { default as postProcessDocumentation } from './postProcessDocumentation';
export { default as printValue } from './printValue';
export { default as resolveExportDeclaration } from './resolveExportDeclaration';
export { default as resolveFunctionDefinitionToReturnValue } from './resolveFunctionDefinitionToReturnValue';
export { default as resolveGenericTypeAnnotation } from './resolveGenericTypeAnnotation';
export { default as resolveHOC } from './resolveHOC';
export { default as resolveObjectPatternPropertyToValue } from './resolveObjectPatternPropertyToValue';
export {
  default as resolveObjectKeysToArray,
  resolveObjectToNameArray,
} from './resolveObjectKeysToArray';
export { default as resolveObjectValuesToArray } from './resolveObjectValuesToArray';
export { default as resolveToModule } from './resolveToModule';
export { default as resolveToValue } from './resolveToValue';
export { default as setPropDescription } from './setPropDescription';
