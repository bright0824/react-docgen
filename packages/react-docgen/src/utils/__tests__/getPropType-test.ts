import type { ExpressionStatement } from '@babel/types';
import { parse, makeMockImporter } from '../../../tests/utils';
import getPropType from '../getPropType.js';
import { describe, expect, test } from 'vitest';
import type { NodePath } from '@babel/traverse';

describe('getPropType', () => {
  test('detects simple prop types', () => {
    const simplePropTypes = [
      'array',
      'bool',
      'func',
      'number',
      'object',
      'string',
      'any',
      'element',
      'node',
      'symbol',
      'elementType',
    ];

    simplePropTypes.forEach((type) =>
      expect(getPropType(parse.expression('React.PropTypes.' + type))).toEqual({
        name: type,
      }),
    );

    // It doesn't actually matter what the MemberExpression is
    simplePropTypes.forEach((type) =>
      expect(getPropType(parse.expression('Foo.' + type + '.bar'))).toEqual({
        name: type,
      }),
    );

    // Doesn't even have to be a MemberExpression
    simplePropTypes.forEach((type) =>
      expect(getPropType(parse.expression(type))).toEqual({
        name: type,
      }),
    );
  });

  test('detects complex prop types', () => {
    expect(getPropType(parse.expression('oneOf(["foo", "bar"])'))).toEqual({
      name: 'enum',
      value: [
        { value: '"foo"', computed: false },
        { value: '"bar"', computed: false },
      ],
    });

    // line comments are ignored
    expect(
      getPropType(parse.expression('oneOf(["foo", // baz\n"bar"])')),
    ).toEqual({
      name: 'enum',
      value: [
        { value: '"foo"', computed: false },
        { value: '"bar"', computed: false },
      ],
    });

    expect(getPropType(parse.expression('oneOfType([number, bool])'))).toEqual({
      name: 'union',
      value: [{ name: 'number' }, { name: 'bool' }],
    });

    // custom type
    expect(getPropType(parse.expression('oneOfType([foo])'))).toEqual({
      name: 'union',
      value: [{ name: 'custom', raw: 'foo' }],
    });

    expect(getPropType(parse.expression('instanceOf(Foo)'))).toEqual({
      name: 'instanceOf',
      value: 'Foo',
    });

    expect(getPropType(parse.expression('arrayOf(string)'))).toEqual({
      name: 'arrayOf',
      value: { name: 'string' },
    });

    expect(getPropType(parse.expression('objectOf(string)'))).toEqual({
      name: 'objectOf',
      value: { name: 'string' },
    });

    expect(
      getPropType(parse.expression('shape({foo: string, bar: bool})')),
    ).toEqual({
      name: 'shape',
      value: {
        foo: {
          name: 'string',
          required: false,
        },
        bar: {
          name: 'bool',
          required: false,
        },
      },
    });

    expect(
      getPropType(parse.expression('exact({foo: string, bar: bool})')),
    ).toEqual({
      name: 'exact',
      value: {
        foo: {
          name: 'string',
          required: false,
        },
        bar: {
          name: 'bool',
          required: false,
        },
      },
    });

    // custom
    expect(getPropType(parse.expression('shape({foo: xyz})'))).toEqual({
      name: 'shape',
      value: {
        foo: {
          name: 'custom',
          raw: 'xyz',
          required: false,
        },
      },
    });

    // custom
    expect(getPropType(parse.expression('exact({foo: xyz})'))).toEqual({
      name: 'exact',
      value: {
        foo: {
          name: 'custom',
          raw: 'xyz',
          required: false,
        },
      },
    });

    // computed
    expect(getPropType(parse.expression('shape(Child.propTypes)'))).toEqual({
      name: 'shape',
      value: 'Child.propTypes',
      computed: true,
    });

    // computed
    expect(getPropType(parse.expression('exact(Child.propTypes)'))).toEqual({
      name: 'exact',
      value: 'Child.propTypes',
      computed: true,
    });
  });

  describe('resolve identifier to their values', () => {
    const mockImporter = makeMockImporter({
      shape: (stmt) =>
        stmt(`
        export default {bar: PropTypes.string};
      `).get('declaration'),

      types: (stmt) =>
        stmt(`
        export default ["foo", "bar"];
      `).get('declaration'),

      foo: (stmt) =>
        stmt(`
        export default "foo";
      `).get('declaration'),

      bar: (stmt) =>
        stmt(`
        export default "bar";
      `).get('declaration'),

      obj: (stmt) =>
        stmt(`
        export default { FOO: "foo", BAR: "bar" };
      `).get('declaration'),

      arr: (stmt) =>
        stmt(`
        export default ["foo", "bar"];
      `).get('declaration'),

      keys: (stmt) =>
        stmt(`
        import obj from 'obj';
        export default Object.keys(obj);
      `).get('declaration'),

      values: (stmt) =>
        stmt(`
        import obj from 'obj';
        export default Object.values(obj);
      `).get('declaration'),
    });

    test('resolves variables to their values', () => {
      const propTypeExpression = parse
        .statement<ExpressionStatement>(
          `
        PropTypes.shape(shape);
        var shape = {bar: PropTypes.string};
      `,
        )
        .get('expression');

      expect(getPropType(propTypeExpression)).toMatchSnapshot();
    });

    test('resolves imported variables to their values', () => {
      const propTypeExpression = parse
        .statement<ExpressionStatement>(
          `
        PropTypes.shape(shape);
        import shape from 'shape';
      `,
          mockImporter,
        )
        .get('expression');

      expect(getPropType(propTypeExpression)).toMatchSnapshot();
    });

    test('resolves simple identifier to their initialization value', () => {
      const propTypeIdentifier = parse
        .statement<ExpressionStatement>(
          `
        PropTypes.oneOf(TYPES);
        var TYPES = ["foo", "bar"];
      `,
        )
        .get('expression');

      expect(getPropType(propTypeIdentifier)).toMatchSnapshot();
    });

    test('resolves importer identifier to initialization value', () => {
      const propTypeIdentifier = parse
        .statement<ExpressionStatement>(
          `
        PropTypes.oneOf(TYPES);
        import TYPES from 'types';
      `,
          mockImporter,
        )
        .get('expression');

      expect(getPropType(propTypeIdentifier)).toMatchSnapshot();
    });

    test('resolves simple identifier to their initialization value in array', () => {
      const identifierInsideArray = parse
        .statement<ExpressionStatement>(
          `
        PropTypes.oneOf([FOO, BAR]);
        var FOO = "foo";
        var BAR = "bar";
      `,
        )
        .get('expression');

      expect(getPropType(identifierInsideArray)).toMatchSnapshot();
    });

    test('resolves imported identifier to their initialization value in array', () => {
      const identifierInsideArray = parse
        .statement<ExpressionStatement>(
          `
        PropTypes.oneOf([FOO, BAR]);
        import FOO from 'foo';
        import BAR from 'bar';
      `,
          mockImporter,
        )
        .get('expression');

      expect(getPropType(identifierInsideArray)).toMatchSnapshot();
    });

    test('handles unresolved imported identifier to their initialization value in array', () => {
      const identifierInsideArray = parse
        .statement<ExpressionStatement>(
          `
        PropTypes.oneOf([FOO, BAR]);
        import FOO from 'foo';
        import BAR from 'bar';
      `,
        )
        .get('expression');

      expect(getPropType(identifierInsideArray)).toMatchSnapshot();
    });

    test('handles unresolved named imported identifier to their initialization value in array', () => {
      const identifierInsideArray = parse
        .statement<ExpressionStatement>(
          `
        PropTypes.oneOf([FOO, BAR]);
        import { FOO } from 'foo';
        import { BAR } from 'bar';
      `,
        )
        .get('expression');

      expect(getPropType(identifierInsideArray)).toMatchSnapshot();
    });

    test('resolves memberExpressions', () => {
      const propTypeExpression = parse
        .statement<ExpressionStatement>(
          `
        PropTypes.oneOf([TYPES.FOO, TYPES.BAR]);
        var TYPES = { FOO: "foo", BAR: "bar" };
      `,
        )
        .get('expression');

      expect(getPropType(propTypeExpression)).toMatchSnapshot();
    });

    test('resolves memberExpressions from imported objects', () => {
      const propTypeExpression = parse
        .statement<ExpressionStatement>(
          `
        PropTypes.oneOf([TYPES.FOO, TYPES.BAR]);
        import TYPES from 'obj';
      `,
          mockImporter,
        )
        .get('expression');

      expect(getPropType(propTypeExpression)).toMatchSnapshot();
    });

    test('correctly resolves SpreadElements in arrays', () => {
      const propTypeExpression = parse
        .statement<ExpressionStatement>(
          `
        PropTypes.oneOf([...TYPES]);
        var TYPES = ["foo", "bar"];
      `,
        )
        .get('expression');

      expect(getPropType(propTypeExpression)).toMatchSnapshot();
    });

    test('correctly resolves SpreadElements in arrays from imported values', () => {
      const propTypeExpression = parse
        .statement<ExpressionStatement>(
          `
        PropTypes.oneOf([...TYPES]);
        import TYPES from 'arr';
      `,
          mockImporter,
        )
        .get('expression');

      expect(getPropType(propTypeExpression)).toMatchSnapshot();
    });

    test('correctly resolves nested SpreadElements in arrays', () => {
      const propTypeExpression = parse
        .statement<ExpressionStatement>(
          `
        PropTypes.oneOf([...TYPES]);
        var TYPES = ["foo", ...TYPES2];
        var TYPES2 = ["bar"];
      `,
        )
        .get('expression');

      expect(getPropType(propTypeExpression)).toMatchSnapshot();
    });

    test('does resolve object keys values', () => {
      const propTypeExpression = parse
        .statement<ExpressionStatement>(
          `
        PropTypes.oneOf(Object.keys(TYPES));
        var TYPES = { FOO: "foo", BAR: "bar" };
      `,
        )
        .get('expression');

      expect(getPropType(propTypeExpression)).toMatchSnapshot();
    });

    test('resolves values from imported Object.keys call', () => {
      const propTypeExpression = parse
        .statement<ExpressionStatement>(
          `
        PropTypes.oneOf(keys);
        import keys from 'keys';
      `,
          mockImporter,
        )
        .get('expression');

      expect(getPropType(propTypeExpression)).toMatchSnapshot();
    });

    test('does resolve object values', () => {
      const propTypeExpression = parse
        .statement<ExpressionStatement>(
          `
        PropTypes.oneOf(Object.values(TYPES));
        var TYPES = { FOO: "foo", BAR: "bar" };
      `,
        )
        .get('expression');

      expect(getPropType(propTypeExpression)).toMatchSnapshot();
    });

    test('resolves values from imported Object.values call', () => {
      const propTypeExpression = parse
        .statement<ExpressionStatement>(
          `
        PropTypes.oneOf(values);
        import values from 'values';
      `,
          mockImporter,
        )
        .get('expression');

      expect(getPropType(propTypeExpression)).toMatchSnapshot();
    });

    test('does not resolve external values without proper importer', () => {
      const propTypeExpression = parse
        .statement<ExpressionStatement>(
          `
        PropTypes.oneOf(TYPES);
        import { TYPES } from './foo.js';
      `,
        )
        .get('expression');

      expect(getPropType(propTypeExpression)).toMatchSnapshot();
    });
  });

  test('detects custom validation functions for function', () => {
    expect(getPropType(parse.expression('(function() {})'))).toMatchSnapshot();
  });

  test('detects custom validation functions for arrow function', () => {
    expect(getPropType(parse.expression('() => {}'))).toMatchSnapshot();
  });

  test('detects descriptions on nested types in arrayOf', () => {
    expect(
      getPropType(
        parse.expression(`arrayOf(
      /**
       * test2
       */
      string
    )`),
      ),
    ).toMatchSnapshot();
  });

  test('detects descriptions on nested types in objectOf', () => {
    expect(
      getPropType(
        parse.expression(`objectOf(
      /**
       * test2
       */
      string
    )`),
      ),
    ).toMatchSnapshot();
  });

  test('detects descriptions on nested types in shapes', () => {
    expect(
      getPropType(
        parse.expression(`shape({
      /**
       * test1
       */
      foo: string,
      /**
       * test2
       */
      bar: bool
    })`),
      ),
    ).toMatchSnapshot();
  });

  test('detects required notations of nested types in shapes', () => {
    expect(
      getPropType(
        parse.expression(`shape({
      foo: string.isRequired,
      bar: bool
    })`),
      ),
    ).toMatchSnapshot();
  });

  test('detects descriptions on nested types in exacts', () => {
    expect(
      getPropType(
        parse.expression(`exact({
      /**
       * test1
       */
      foo: string,
      /**
       * test2
       */
      bar: bool
    })`),
      ),
    ).toMatchSnapshot();
  });

  test('detects required notations of nested types in exacts', () => {
    expect(
      getPropType(
        parse.expression(`exact({
      foo: string.isRequired,
      bar: bool
    })`),
      ),
    ).toMatchSnapshot();
  });

  test('handles computed properties', () => {
    expect(
      getPropType(
        parse.expression(`exact({
      [foo]: string.isRequired,
      bar: bool
    })`),
      ),
    ).toMatchSnapshot();
  });

  test('ignores complex computed properties', () => {
    expect(
      getPropType(
        parse.expression(`exact({
      [() => {}]: string.isRequired,
      bar: bool
    })`),
      ),
    ).toMatchSnapshot();
  });

  test('works with cyclic references in shape', () => {
    expect(
      getPropType(
        parse
          .statementLast<ExpressionStatement>(
            `const Component = () => {}
             Component.propTypes = {
               foo: shape(Component.propTypes)
             }`,
          )
          .get('expression.right.properties.0.value') as NodePath,
      ),
    ).toMatchSnapshot();
  });

  test('works with cyclic references in shape and required', () => {
    expect(
      getPropType(
        parse
          .statementLast<ExpressionStatement>(
            `const Component = () => {}
             Component.propTypes = {
               foo: shape(Component.propTypes).isRequired
             }`,
          )
          .get('expression.right.properties.0.value') as NodePath,
      ),
    ).toMatchSnapshot();
  });

  test('works with missing argument', () => {
    expect(
      getPropType(
        parse
          .statementLast<ExpressionStatement>(
            `const Component = () => {}
             const MyShape = { foo: shape() }
             Component.propTypes = {
               foo: shape(MyShape)
             }`,
          )
          .get('expression.right.properties.0.value') as NodePath,
      ),
    ).toMatchSnapshot();
  });
});
