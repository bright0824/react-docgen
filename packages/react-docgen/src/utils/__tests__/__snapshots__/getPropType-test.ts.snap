// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getPropType > detects custom validation functions for arrow function 1`] = `
{
  "name": "custom",
  "raw": "() => {}",
}
`;

exports[`getPropType > detects custom validation functions for function 1`] = `
{
  "name": "custom",
  "raw": "function() {}",
}
`;

exports[`getPropType > detects descriptions on nested types in arrayOf 1`] = `
{
  "description": "test2",
  "name": "arrayOf",
  "value": {
    "name": "string",
  },
}
`;

exports[`getPropType > detects descriptions on nested types in exacts 1`] = `
{
  "name": "exact",
  "value": {
    "bar": {
      "description": "test2",
      "name": "bool",
      "required": false,
    },
    "foo": {
      "description": "test1",
      "name": "string",
      "required": false,
    },
  },
}
`;

exports[`getPropType > detects descriptions on nested types in objectOf 1`] = `
{
  "description": "test2",
  "name": "objectOf",
  "value": {
    "name": "string",
  },
}
`;

exports[`getPropType > detects descriptions on nested types in shapes 1`] = `
{
  "name": "shape",
  "value": {
    "bar": {
      "description": "test2",
      "name": "bool",
      "required": false,
    },
    "foo": {
      "description": "test1",
      "name": "string",
      "required": false,
    },
  },
}
`;

exports[`getPropType > detects required notations of nested types in exacts 1`] = `
{
  "name": "exact",
  "value": {
    "bar": {
      "name": "bool",
      "required": false,
    },
    "foo": {
      "name": "string",
      "required": true,
    },
  },
}
`;

exports[`getPropType > detects required notations of nested types in shapes 1`] = `
{
  "name": "shape",
  "value": {
    "bar": {
      "name": "bool",
      "required": false,
    },
    "foo": {
      "name": "string",
      "required": true,
    },
  },
}
`;

exports[`getPropType > handles computed properties 1`] = `
{
  "name": "exact",
  "value": {
    "@computed#foo": {
      "name": "string",
      "required": true,
    },
    "bar": {
      "name": "bool",
      "required": false,
    },
  },
}
`;

exports[`getPropType > ignores complex computed properties 1`] = `
{
  "name": "exact",
  "value": {
    "bar": {
      "name": "bool",
      "required": false,
    },
  },
}
`;

exports[`getPropType > resolve identifier to their values > correctly resolves SpreadElements in arrays 1`] = `
{
  "name": "enum",
  "value": [
    {
      "computed": false,
      "value": "\\"foo\\"",
    },
    {
      "computed": false,
      "value": "\\"bar\\"",
    },
  ],
}
`;

exports[`getPropType > resolve identifier to their values > correctly resolves SpreadElements in arrays from imported values 1`] = `
{
  "name": "enum",
  "value": [
    {
      "computed": false,
      "value": "\\"foo\\"",
    },
    {
      "computed": false,
      "value": "\\"bar\\"",
    },
  ],
}
`;

exports[`getPropType > resolve identifier to their values > correctly resolves nested SpreadElements in arrays 1`] = `
{
  "name": "enum",
  "value": [
    {
      "computed": false,
      "value": "\\"foo\\"",
    },
    {
      "computed": false,
      "value": "\\"bar\\"",
    },
  ],
}
`;

exports[`getPropType > resolve identifier to their values > does not resolve external values without proper importer 1`] = `
{
  "computed": true,
  "name": "enum",
  "value": "TYPES",
}
`;

exports[`getPropType > resolve identifier to their values > does resolve object keys values 1`] = `
{
  "name": "enum",
  "value": [
    {
      "computed": false,
      "value": "\\"FOO\\"",
    },
    {
      "computed": false,
      "value": "\\"BAR\\"",
    },
  ],
}
`;

exports[`getPropType > resolve identifier to their values > does resolve object values 1`] = `
{
  "name": "enum",
  "value": [
    {
      "computed": false,
      "value": "\\"foo\\"",
    },
    {
      "computed": false,
      "value": "\\"bar\\"",
    },
  ],
}
`;

exports[`getPropType > resolve identifier to their values > handles unresolved imported identifier to their initialization value in array 1`] = `
{
  "name": "enum",
  "value": [
    {
      "computed": true,
      "value": "FOO",
    },
    {
      "computed": true,
      "value": "BAR",
    },
  ],
}
`;

exports[`getPropType > resolve identifier to their values > handles unresolved named imported identifier to their initialization value in array 1`] = `
{
  "name": "enum",
  "value": [
    {
      "computed": true,
      "value": "FOO",
    },
    {
      "computed": true,
      "value": "BAR",
    },
  ],
}
`;

exports[`getPropType > resolve identifier to their values > resolves imported identifier to their initialization value in array 1`] = `
{
  "name": "enum",
  "value": [
    {
      "computed": false,
      "value": "\\"foo\\"",
    },
    {
      "computed": false,
      "value": "\\"bar\\"",
    },
  ],
}
`;

exports[`getPropType > resolve identifier to their values > resolves imported variables to their values 1`] = `
{
  "name": "shape",
  "value": {
    "bar": {
      "name": "string",
      "required": false,
    },
  },
}
`;

exports[`getPropType > resolve identifier to their values > resolves importer identifier to initialization value 1`] = `
{
  "name": "enum",
  "value": [
    {
      "computed": false,
      "value": "\\"foo\\"",
    },
    {
      "computed": false,
      "value": "\\"bar\\"",
    },
  ],
}
`;

exports[`getPropType > resolve identifier to their values > resolves memberExpressions 1`] = `
{
  "name": "enum",
  "value": [
    {
      "computed": false,
      "value": "\\"foo\\"",
    },
    {
      "computed": false,
      "value": "\\"bar\\"",
    },
  ],
}
`;

exports[`getPropType > resolve identifier to their values > resolves memberExpressions from imported objects 1`] = `
{
  "name": "enum",
  "value": [
    {
      "computed": false,
      "value": "\\"foo\\"",
    },
    {
      "computed": false,
      "value": "\\"bar\\"",
    },
  ],
}
`;

exports[`getPropType > resolve identifier to their values > resolves simple identifier to their initialization value 1`] = `
{
  "name": "enum",
  "value": [
    {
      "computed": false,
      "value": "\\"foo\\"",
    },
    {
      "computed": false,
      "value": "\\"bar\\"",
    },
  ],
}
`;

exports[`getPropType > resolve identifier to their values > resolves simple identifier to their initialization value in array 1`] = `
{
  "name": "enum",
  "value": [
    {
      "computed": false,
      "value": "\\"foo\\"",
    },
    {
      "computed": false,
      "value": "\\"bar\\"",
    },
  ],
}
`;

exports[`getPropType > resolve identifier to their values > resolves values from imported Object.keys call 1`] = `
{
  "name": "enum",
  "value": [
    {
      "computed": false,
      "value": "\\"FOO\\"",
    },
    {
      "computed": false,
      "value": "\\"BAR\\"",
    },
  ],
}
`;

exports[`getPropType > resolve identifier to their values > resolves values from imported Object.values call 1`] = `
{
  "name": "enum",
  "value": [
    {
      "computed": false,
      "value": "\\"foo\\"",
    },
    {
      "computed": false,
      "value": "\\"bar\\"",
    },
  ],
}
`;

exports[`getPropType > resolve identifier to their values > resolves variables to their values 1`] = `
{
  "name": "shape",
  "value": {
    "bar": {
      "name": "string",
      "required": false,
    },
  },
}
`;

exports[`getPropType > works with cyclic references in shape 1`] = `
{
  "name": "shape",
  "value": "Component.propTypes",
}
`;

exports[`getPropType > works with cyclic references in shape and required 1`] = `
{
  "name": "shape",
  "value": "Component.propTypes",
}
`;

exports[`getPropType > works with missing argument 1`] = `
{
  "name": "shape",
  "value": {
    "foo": {
      "computed": true,
      "name": "shape",
      "required": false,
      "value": "",
    },
  },
}
`;
